File: cpp/trex_selector_methods/trex_da/demo_trex_da_06_mc_sim_groups/demo_trex_da_06_mc_sim_groups.cpp
// ==============================================================================
// demo_trex_da_06_mc_sim_groups.cpp
// ==============================================================================
/**
 * @file demo_trex_da_06_mc_sim_groups.cpp
 *
 * @brief DA-TRex MC simulation for the prior-groups DGP.
 *
 * @details
 *  Functions:
 *   demo_groups_mc_snr_sweep() -- SNR sweep (n=300, p=1000, s=10, 200 MC, 3 solvers).
 */
// ==============================================================================

#include "trex_da_sim_common.hpp"

using namespace da_sim;


// ==============================================================================
// Group structure builder
// ==============================================================================

/**
 * @brief Build a 3-level hierarchical group structure for p predictors.
 *
 * Level 1: groups of 10   (p/10 groups)
 * Level 2: groups of 50   (p/50 groups)
 * Level 3: groups of 250  (p/250 groups)
 */
std::vector<std::vector<Eigen::Index>> make_group_structure(int p) {
    std::vector<std::vector<Eigen::Index>> groups(3);
    for (Eigen::Index j = 0; j < p; ++j) {
        groups[0].push_back(j / 10);
        groups[1].push_back(j / 50);
        groups[2].push_back(j / 250);
    }
    return groups;
}

namespace {
    constexpr int max_gap = 20;
}


// ==============================================================================
//  MC SNR sweep
//  DGP: dgp_groups_toeplitz_leaf() with 3-level hierarchy, Random support.
//  Fixed: n=300, p=1000, s=10 (SimConfig defaults), tFDR=0.2, K=20, num_MC=200.
//  Swept: SNR in {0.1, 0.2, 0.5, 1.0, 2.0, 5.0}.
// ==============================================================================
void demo_groups_mc_snr_sweep()
{
    SimConfig cfg;  // n=300, p=1000, s=10
    const auto policy = SupportPolicy::Random;

    auto groups = make_group_structure(cfg.p);

     // cumulative variance masses, fine -> coarse
    const std::vector<double> rho_levels = {0.55, 0.25, 0.10};
    constexpr double phi_leaf = 0.50;

    // The prior groups act as merge CONSTRAINTS: the selector sub-clusters
    // within their finest common refinement (the groups of 10) and calibrates
    // over a data-driven rho grid ending at the conservative rho = 1 anchor.
    TRexDAControlParameter da_ctrl;
    da_ctrl.prior_groups = groups;

    run_snr_sweep(
        "da_groups_toeplitz_leaf",
        {0.1, 0.2, 0.5, 1.0, 2.0, 5.0},
        cfg.num_MC,
        cfg.tFDR,
        cfg.base_seed,
        default_solvers(),
        [&](double snr, unsigned seed) {
            return dgp_groups_toeplitz_leaf(
                cfg.n,
                cfg.p,
                make_support(policy, cfg.s, cfg.p, max_gap, seed),
                cfg.amplitude,
                groups,
                rho_levels,
                phi_leaf,
                snr,
                seed
            );
        },
        da_ctrl,
        make_trex_control(cfg.K),
        "Prior Groups + Toeplitz leaf, n=" + std::to_string(cfg.n)
        + ", p=" + std::to_string(cfg.p)
        + ", support=" + support_policy_label(policy)
        + ", phi_leaf=" + fmt_num(phi_leaf)
        + ", max_gap=" + std::to_string(max_gap));
}


// ==============================================================================
// main
// ==============================================================================
int main() {
    std::cout.setf(std::ios::unitbuf);
    omp_set_num_threads(6);
    std::cout << "Running with " << omp_get_max_threads() << " threads\n\n";

    if (true) demo_groups_mc_snr_sweep();

    std::cout << "\nPrior Groups MC simulations complete.\n";
    return 0;
}

File: Python/trex_selector_methods/trex_da/demo_trex_da_06_mc_sim_groups/demo_trex_da_06_mc_sim_groups.py
# ==============================================================================
# demo_trex_da_06_mc_sim_groups.py
# ==============================================================================
#
# DA-TRex MC simulation for the prior-groups DGP.
#
# The predictors carry a known 3-level hierarchical group structure (groups of
# 10 / 50 / 250, fine -> coarse), passed to the selector via
# da_control.prior_groups. The PRIOR_GROUPS method (redefined 2026-07-17)
# treats these labels as merge CONSTRAINTS: the selector sub-clusters within
# their finest common refinement (hc_linkage, correlation distance) and
# calibrates over a data-driven ascending rho grid terminated by the
# conservative rho = 1 singleton anchor. Data is drawn from a multi-level
# latent-factor model with Toeplitz leaf blocks.
#
#   Part 1: SNR sweep (n=300, p=1000, s=10, Random support, 200 MC, 3 solvers).
#
# Mirrors cpp/trex_selector_methods/trex_da/demo_trex_da_06_mc_sim_groups
# (same DGP, control parameters, sweep axis, and result files). Results are
# saved to simulation_results/ as TXT and CSV files. Reduce NUM_MC to preview.
# ==============================================================================

import os
import sys

_THIS_DIR = os.path.dirname(os.path.abspath(__file__))
_PARENT_DIR = os.path.dirname(_THIS_DIR)
for _p in (_THIS_DIR, _PARENT_DIR):
    if _p not in sys.path:
        sys.path.insert(0, _p)

from da_sim_common import default_solvers, fmt_num, run_snr_sweep

RUN_PART_1 = True

OUT_DIR = os.path.join(_THIS_DIR, "simulation_results")

# Simulation config (mirrors cpp SimConfig defaults)
N, P, S = 300, 1000, 10
AMPLITUDE = 3.0
TFDR = 0.2
K = 20
NUM_MC = 200
BASE_SEED = 2026
MAX_GAP = 20   # unused by the Random policy; kept for cpp parity

# 3-level hierarchy: groups of 10, 50, 250 (fine -> coarse) and their
# cumulative variance masses, fine -> coarse.
RHO_LEVELS = [0.55, 0.25, 0.10]
PHI_LEAF = 0.50


def make_group_structure(p):
    """Level 1: groups of 10; level 2: groups of 50; level 3: groups of 250."""
    return [[j // 10 for j in range(p)],
            [j // 50 for j in range(p)],
            [j // 250 for j in range(p)]]


# ==============================================================================
# Part 1: MC SNR sweep
# DGP: dgp_groups_toeplitz_leaf with 3-level hierarchy, Random support.
# Fixed: n=300, p=1000, s=10, tFDR=0.2, K=20, num_MC=200.
# Swept: SNR in {0.1, 0.2, 0.5, 1.0, 2.0, 5.0}.
# ==============================================================================

def part_1_snr_sweep():
    groups = make_group_structure(P)

    # The prior groups act as merge CONSTRAINTS: the selector sub-clusters
    # within their finest common refinement (the groups of 10) and calibrates
    # over a data-driven rho grid ending at the conservative rho = 1 anchor.
    da_spec = {"prior_groups": groups}

    run_snr_sweep(
        OUT_DIR, "da_groups_toeplitz_leaf",
        [0.1, 0.2, 0.5, 1.0, 2.0, 5.0], NUM_MC, TFDR, BASE_SEED,
        default_solvers(),
        lambda snr: dict(n=N, p=P, dgp="dgp_groups_toeplitz_leaf",
                         kwargs=dict(amplitude=AMPLITUDE, groups=groups,
                                     rho_levels=RHO_LEVELS, phi=PHI_LEAF,
                                     snr=snr),
                         support=None, s=S),
        da_spec, K,
        f"Prior Groups + Toeplitz leaf, n={N}, p={P}, support=Random"
        f", phi_leaf={fmt_num(PHI_LEAF)}, max_gap={MAX_GAP}")


# ==============================================================================
# Main
# ==============================================================================

if __name__ == "__main__":
    if RUN_PART_1:
        part_1_snr_sweep()
    print("\nPrior Groups MC simulations complete.")
